lib/task-queue/TaskRouter.js

```diff
@@ -0,0 +1,133 @@
+/* globals Task */
+"use strict";
+
+const {Cu} = require("chrome");
+const {WorkerTask} = require("lib/task-queue/WorkerTask");
+const {Storage} = require("lib/task-queue/Storage");
+const {Page} = require("lib/task-queue/Page");
+
+Cu.import("resource://gre/modules/Task.jsm");
+
+// The lock for page and task creation test-and-set
+let taskCreationLock = false;
+// The queue for messages waiting to be processed
+const mutexQueue = [];
+
+class TaskRouter {
+  /**
+   * Creates a new task router from a map of messages to TaskProcessor arrays
+   *
+   * @param {Map<string, TaskProcessor[]>} processorMap
+   *                                       A map of Task Processors arrays for each message
+   */
+  constructor(processorMap) {
+    if (!processorMap) {
+      throw new Error("TaskRouter needs a map as argument.");
+    }
+    // Initialize task processors
+    this.routes = processorMap;
+  }
+
+  /**
+   * Check if the page should be processed by the queue
+   *
+   * @param {String}    url
+   *                    The URL of the page we want to check
+   * @returns {Promise}
+   *                    Returns true if the page should be processed, false otherwise
+   */
+  asyncNeedProcessing(url) {
+    return Task.spawn((function*() {
+      // Do we already have a task for this url?
+      const tasks = yield WorkerTask.asyncGetByUrl(url);
+      if (tasks) {
+        return false;
+      }
+
+      // Do we already have a page for this url?
+      const page = yield Page.asyncGetByUrl(url);
```

Comment:
Since we don't have page paths anymore, we check only the page expiration here.
What do you think about a query that would get only the expiration? Would the performance gains be significant enough?

Something like:
```sql
SELECT (ISNULL(createdAt,0) + ISNULL(maxAge,0)) as 'expiration' FROM moz_pages WHERE url = :url
```